### ISSUE TITLE
SelectStyle requested within looplet nest and Out-of-order writes do not error when not supported

### ISSUE THREAD

**Issue by kylebd99** [willow-ahrens/Finch.jl#115]
I'm running into an error when trying to use a custom operator. Further, registering an unchanged algebra results in a different error, and registering an algebra which references the custom operator produces a different error. As a side note, before I updated from 0.2.0 to the current main branch, the second and third code samples produced the correct output.

For example, when I run this code:
```
function f(a, b, c)
    return a+b+c
end 
t = @fiber(sl(sl(e(0.0))))
A = dropdefaults(copyto!(@fiber(sl(sl(e(0.0)))), SparseMatrixCSC([0 0 0 0; -1 -1 -1 -1; -2 -2 -2 -2; -3 -3 -3 -3])))
@finch @loop i j t[i, j] = f(A[i,j], A[i,j], A[i,j])
```
I get the following error:
```
RewriteTools.RuleRewriteError(call(~f, ~a...) => if isliteral(f) && (all(isliteral, a) && length(a) >= 1)
        #= /home/kylebd99/.julia/packages/Finch/De0AF/src/annihilate.jl:155 =#
        literal((getvalue(f))(getvalue.(a)...))
    end, Finch.IndexNotation.IndexNode(Finch.IndexNotation.call, nothing, nothing, Finch.IndexNotation.IndexNode[Finch.IndexNotation.IndexNode(Finch.IndexNotation.literal, f, nothing, Finch.IndexNotation.IndexNode[]), Finch.IndexNotation.IndexNode(Finch.IndexNotation.literal, 0.0, nothing, Finch.IndexNotation.IndexNode[]), Finch.IndexNotation.IndexNode(Finch.IndexNotation.literal, 0.0, nothing, Finch.IndexNotation.IndexNode[]), Finch.IndexNotation.IndexNode(Finch.IndexNotation.literal, 0.0, nothing, Finch.IndexNotation.IndexNode[])]))
```
However, if I run the following code:
```
function f(a::Float64, b::Float64, c::Float64)
    return a+b+c
end 
struct MyAlgebra <: Finch.AbstractAlgebra end
Finch.register(MyAlgebra)
t = @fiber(sl(sl(e(0.0))))
A = dropdefaults(copyto!(@fiber(sl(sl(e(0.0)))), SparseMatrixCSC([0 0 0 0; -1 -1 -1 -1; -2 -2 -2 -2; -3 -3 -3 -3])))
@finch MyAlgebra() @loop i j t[i, j] = f(A[i,j], A[i,j], A[i,j])
```
I get the following error:
```
MethodError: no method matching combine_style(::Finch.RunStyle, ::Finch.SelectStyle)
Closest candidates are:
  combine_style(::Any, ::Any) at ~/.julia/packages/Finch/NfCmw/src/style.jl:40
  combine_style(::Finch.RunStyle, ::Finch.RunStyle) at ~/.julia/packages/Finch/NfCmw/src/looplets/runs.jl:22
  combine_style(::Finch.RunStyle, ::Finch.AcceptRunStyle) at ~/.julia/packages/Finch/NfCmw/src/looplets/runs.jl:97
```
Lastly, if I run the following code:
```
function f(a, b, c)
    return a+b+c
end 
struct MyAlgebra2 <: Finch.AbstractAlgebra end
Finch.isassociative(::MyAlgebra2, ::typeof(f)) = true
Finch.iscommutative(::MyAlgebra2, ::typeof(f)) = true
Finch.register(MyAlgebra2)
t = @fiber(sl(sl(e(0.0))))
A = dropdefaults(copyto!(@fiber(sl(sl(e(0.0)))), SparseMatrixCSC([0 0 0 0; -1 -1 -1 -1; -2 -2 -2 -2; -3 -3 -3 -3])))
@finch MyAlgebra2() @loop i j t[i, j] = f(A[i,j], A[i,j], A[i,j])
```
I get the following error:
```
RewriteTools.RuleRewriteError(call(~(f::isassociative(alg)), ~a..., ~(b::isliteral), ~(c::isliteral), ~d...) => call(f, a..., f.val(b.val, c.val), d...), call(literal(f), access(virtual(Finch.Furlable(nothing, nothing, (Finch.Extent(literal(1), value(A_lvl_2.I, Int64), call(literal(+), call(literal(-), value(A_lvl_2.I, Int64), literal(1)), literal(1)), call(literal(+), call(literal(-), value(A_lvl_2.I, Int64), literal(1)), literal(1))),), Finch.var"#398#405"{Tuple{}, Symbol, Symbol, Symbol, Symbol, DataType, DataType, Finch.FinchNotation.FinchNode, Finch.VirtualSparseListLevel}((), :A_lvl_2_i1, :A_lvl_2_q_stop, :A_lvl_2_q, :A_lvl_2_i, Int64, Int64, value(A_lvl_q_2, Int64), Finch.VirtualSparseListLevel(Finch.VirtualElementLevel(:A_lvl_3, Float64, 0.0), :A_lvl_2, Int64, Int64, value(A_lvl_2.I, Int64), :A_lvl_2_qos_fill, :A_lvl_2_qos_stop)))), reader(), index(s)), literal(0.0), literal(0.0)))
```

**Comment by willow-ahrens** [comment]
A few observations. First, main is column major and v0.2.0 is row major. Since your examples are row major, when you made the switch Finch is now trying to access (and write to) those arrays out of order. Finch supports random access and out of order access using the SelectVisitor, so that's probably why the error involving SelectStyle showed up after switching to main. Regardless, one should be able to read out of order, and I think I see what's wrong there so I'll push a fix. One cannot write to sparse list levels out of order, but there should be an error for that if the format doesn't support it. 

Regarding the custom rules, I'll investigate a little more and get back to you, but I think it's just that the algebra registration thing isn't watertight and somewhere there's a world age issue that needs a quick patch.

**Comment by willow-ahrens** [comment]
The third error is interesting. Finch thinks that associative operators are also `nary`(and therefore groups them into a flattened representation), and it noticed that two of the arguments to `f` were literals, so it tried to call `f(b, c)`, but no such method is defined.

**Comment by willow-ahrens** [comment]
For the third error, check https://github.com/willow-ahrens/Finch.jl/blob/fa53a1ccc7fb37bcdd9d53e9745e30256a073217/src/annihilate.jl#L19

**Comment by willow-ahrens** [comment]
Fixed by #116
